### DISCUSSION TITLE
Any plans to include data standardisation module?

### DISCUSSION THREAD

**Discussion by amgfernandes** [moj-analytical-services/splink#986] [Ideas]
First of all thank you very much for the great work and providing Splink as open-source. Its a great tool and I used it regularly for several projects. 

Did you consider including a preprocessing module so datasets can be standardised (same column names, and consistent formatting like lowercased, punctuation cleaned up)? 

Something similar to this: https://recordlinkage.readthedocs.io/en/latest/ref-preprocessing.html

Thank you,
Miguel

**Comment by aalexandersson** [comment]
I think it is a great idea. The Chicago-based company DataMade has two useful Python libraries for data standardisation:

[probablepeople](https://github.com/datamade/probablepeople) for parsing names, and [usaddress](https://github.com/datamade/usaddress) for parsing U.S. addresses.

At minimum, I would like to see some `splink` documentation on how to use existing tools such as `probablepeople` and `usaddress`. It would be a great addition to the [data prerequisite tutorial](https://moj-analytical-services.github.io/splink/demos/01_Prerequisites.html).

**Reply by RobinL** [reply]
Nice - wasn't aware of these two libs.  For others reading this, for international addresses, there's also [libpostal](https://github.com/openvenues/libpostal)

**Reply by checkbook-org** [reply]
My experience with libpostal for US addresses is not great - it seemed to have trouble with addresses that didn't seem like they should be problematic. Also, if I'm thinking of the same library, on Windows it required building an multi-GB .dll that seems to contain training data. 

Another couple libraries I came across are:
Deepparse
https://github.com/GRAAL-Research/deepparse

Postmastr in R, not sure if that's US-centric or not.  One benefit of postmastr is that it seems to take a narrowing approach to parsing, so that it lists of cities can be used rather than putting the entire job of recognizing city names on the parser, etc...

**Comment by RobinL** [comment]
There are no current plans to do this, mostly due to resource constraints.  In an ideal world we'd have a robust set of functions for data cleaning, but they're somewhat challenging to write given that we're attempting to support multiple backends.  

I agree that we should provide better docs - e.g. signposting existing libraries

EDIT:  Just for clarification after reading through this thread:

**As it stands, it's unlikely standardisation functions will ever be built into Splink itself.**  Whilst they are undoubtably useful, we want to  avoid bloat and separate concerns.  So if we ever were to work on this, it would be a separate library.

**Reply by amgfernandes** [reply]
Thanks Robin. Thats a great idea about the documentation. I used successfully the Python Record Linkage Toolkit Preprocessing module with Splink so this could maybe be mentioned in the docs.

**Comment by NickCrews** [comment]
I think splink should stay focused and not tackle pre-processing. It can link out to external libraries as suggestions, but I don't think we should try to reinvent the wheel.

**Reply by RobinL** [reply]
Agreed.  Much better to use existing solutions if they exist.

Furthermore, Splink is already a lot of lines of code and I don't want to further increase the maintenance burden.  Also, standardisation functions could be written pretty much independently from Splink, so should probably be a separate library rather than being strongly coupled to Splink itself

**Comment by aflaxman** [comment]
I've been watching this thread with a lot of interest.  Would folks be participating in some brainstorming around preprocessing, and what sort of docs, external libraries, etc exist as well as what gaps there are that splink or other open-source efforts might consider filling?  We could continue collecting ideas on this discussion thread, but it might also be fun to meet in real-time over Zoom or something.

**Reply by NickCrews** [reply]
I've been very happy with [Ibis](https://ibis-project.org/) as the backbone of my data wrangling. It pairs very naturally with splink, is intuitive, very fast, fairly full-featured, and is pluggable enough that I can fill any gaps in features that I need. I then have some fairly custom functions for cleaning names and addresses built on top. Unfortunately I found that most python libs out there work on python objects and aren't fast enough for the 10-50 million rows that I am dealing with (eg usaddress, probablepeople, etc), so I am fairly limited to SQL-based approaches.

**Reply by aalexandersson** [reply]
I think that we should always consider different use cases, also here when discussing any possible plans to include a data standardisation module. What trade-off of speed, accuracy and user-friendliness is most important?

Currently, `splink` with the default `DuckDB` back-end meets all my requirements for speed. But I wish `splink` is more accurate (e.g., could estimate a hit-miss model) and more user-friendly for beginners (e.g., has a GUI).

Specifically for data standardisation, I hope to find use `splink` documentation guidance on these two issues when using the default `DuckDB` back-end:
1) Is there anything better than [https://github.com/datamade/usaddress](https://github.com/datamade/usaddress) for U.S. patient addresses? -- Some time in the future, ideally, it would be great to have an implementation of the new Project US@ standard for U.S. patient addresses: [https://oncprojectracking.healthit.gov/wiki/pages/viewpage.action?pageId=180486153](https://oncprojectracking.healthit.gov/wiki/pages/viewpage.action?pageId=180486153). 
2) How does `splink` best handle partial matches of the 9-digit U.S. Social Security Number (SSN)? Especially, how to handle the case when 7 or 8 digits are correct?  -- Is a comparator function for SSN such as, for example, the [Damerau-Levenshtein edit distance](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance) easily available for `splink`?

**Reply by RobinL** [reply]
At the moment, there's no easy way to write a user-defined function using the DuckDB backend, so you're limited to the functions [native](https://duckdb.org/docs/sql/functions/char.html) to DuckDB.  (It _is_ possible, just hard - e.g. I believe you can write a DuckDB extension).

So in DuckDB right now, you can use Levenshtein easily but not Damerau-Levenshtein.  I would probably just use Levenshtein for now.

With the sqlite and pyspark backends, it's possible to write and register a Python UDF.  These would need to be registered against the sqlite connection and the spark session respectively - so long as they are available when executing SQL you can use them in Splink.

Could you describe what a 'hit-miss' model means?  I've not come across that terminology before.  Whilst Splink models can be inaccurate, we believe that when [well specified models](https://www.robinlinacre.com/comparing_splink_models_unsupervised/) are used, they compare favourably  competing libraries and methodologies. 

We have no current plans to build a GUI for Splink.  Our primary use case is part of automated data engineering pipelines where everything is written in code.  However, we have designed Splink with this possibility in mind.  A GUI could be designed that outputted a Splink settings .json object, so long as it adhered to the [formal jsonschema](https://github.com/moj-analytical-services/splink/blob/master/splink/files/settings_jsonschema.json), which should nicely preserve separation of concerns.

**Reply by NickCrews** [reply]
Re: UDFs in DuckDB, written in python, but callable from SQL, [see my feature request](https://github.com/duckdb/duckdb/discussions/4797)

I am also curious what you mean by a hit-miss model.

**Reply by aalexandersson** [reply]
Thank you so much everyone for the comments to my questions. The **"hit-miss" mode**l is known in the literature (Copas and Hilton 1990, "Record Linkage: Statistical Models for Matching Computer Records" [https://www.jstor.org/stable/2982975](https://www.jstor.org/stable/2982975)) as an example of measuring match error probabilities, not merely match probabilities. I am not sure if extends or rather contrasts with the Fellegi-Sunter model.

Damera-Levenshtein versus Levenshtein is a much simpler, and I hope better, example of showing the potential for more **pre-processing accuracy**  in `splink`. The Damerau-Levenshtein measure allows for transpositions but Levenshtein does not. 

The simplest example of a possible **GUI** for pre-processing `splink` that I can think of is for uploading the input files. A GUI could also be useful for post-processing. Robin, thanks for clarifying the design and plans. A few months ago, you made a comment about possibly developing an interactive dashboard for clerical matching which to me fits in the context of a GUI for post-processing. Is this interactive dashboard still being considered? [https://github.com/moj-analytical-services/splink/discussions/869#discussioncomment-4074861](https://github.com/moj-analytical-services/splink/discussions/869#discussioncomment-4074861)

**Reply by RobinL** [reply]
Thanks for the jstor link - I had not come across this before so I will try and get my head around it!

Re: the GUI for clerical matching.  It's still in development.  It basically works, but frustratingly I just haven't had enough time to finish it off.  For what it's worth, a current mostly-working draft is [here](https://observablehq.com/d/0420fc5db15e44ab).  That link includes all the code and the GUI, but obviously in the final version it would just present the GUI.  But this GUI is totally separate to the Splink code - the only 'interface' between the two is that the input data and output data must be in a specific format.

**Reply by checkbook-org** [reply]
I'm interested in your idea of doing a review of preprocessing tools.
A couple use cases I see as worth some joint development effort are building out more sophisticated comparisons for names and addresses.
1 - Matching Names
Names are challenging because in many cases it's hard to definitively say what is a last name vs a middle name. And of course in non-English speaking cultures there are a variety of ways of handling multiple surnames.
Something that uses the ComparisonLevels component of Splink, and figures out a sensible scheme to parse, clean, and compare names given the uncertainty of whether a part of a name is in the correct field or not, plus good documentation seems like it would address a very common set of use cases and would be reasonably generalizable.
2- Matching addresses using geospatial functions
Along the same lines, some form of address matching that is geographical in nature (thinking of using the new DuckDb geospatial extension), along with good instructions on how to preprocess and use also seems like it would address a common use case.

I have been working with Pelias, and generally it seems like a good tool (North America-centric, if I'm not mistaken).  However, I've had to work around its dependency on libpostal.  Currently investigating usaddress as a replacement. My organization also uses Smarty Streets, and I have been impressed with the data quality.  They recently raised their prices, though, so it's gotten kind of unaffordable.

**Reply by RobinL** [reply]
This is interesting for US addresses:
https://geocoding.geo.census.gov/geocoder/
from the faq:

> The overall goal of the Census Geocoder is to provide the public with the capability to geocode an address efficiently and accurately. There are multiple options that a user can select to geocode an address with the Geocoder. For further information on use of the Census Geocoder, please refer to the User Guide or Frequently Asked Questions documents found below in the Documentation section.

**Reply by RobinL** [reply]
chat gpt also does quite a good job.  this is using the gpt-3.5-turbo model.  would possibly get quite expensive for a large dataset though.  addresses from https://www.mjt.me.uk/posts/falsehoods-programmers-believe-about-addresses/
![image](https://user-images.githubusercontent.com/2608005/224097026-0e2433fa-db13-4d8b-9395-f9489bcbf581.png)

**Reply by aalexandersson** [reply]
This is very interesting! Can `ChatGPT` improve if being told to use a technical specification? For example:

"Standardize the same addresses, but this time the data are for the United States instead of United Kingdom. Use the technical specification of `Project US@` at [https://asapnet.org/wp-content/uploads/2022/03/Project_US_FINAL_Technical_Specification_Version_1.0.pdf](https://asapnet.org/wp-content/uploads/2022/03/Project_US_FINAL_Technical_Specification_Version_1.0.pdf). Use the field name `city` instead of the field `town`. Ignore the fields `village` and `postcode`."

For example, in your example data, the road name "Ammanford Road" should be standardized to "AMMANFORD RD".

**Comment by aalexandersson** [comment]
Re the GUI for clerical matching: I am not sure about the workflow but it seems very promising! Did you consider to prefill the `clerical match score` with the `estimated match probability`? I see two possible advantages of doing it. But maybe some rounding of decimals is needed because I cannot imagine four digits being useful in a clerical review:

1. Speed up the clerical review (at the risk of being biased from pre-filled values). In practice, usually most record pairs do not need clerical review and there is only enough time to clerically review the uncertain record pairs ("possibles").
2. Enable the automatic calculation of a 2-by-2 "summary truth table". It is also known under several other names such as [confusion matrix or error matrix](https://en.wikipedia.org/wiki/Confusion_matrix)". In the R package `fastLink`, the implementation is the function `confusion()`.
